go-controller/pkg/ovn/policy.go

```diff
@@ -685,312 +822,288 @@ func podDeleteAllowMulticastPolicy(nbClient libovsdbclient.Client, ns string, po
+// 1. oc.lsp[In/E]gressDenyCache counters
+// 2. default deny port groups in the db
+// 3. np.localPods
+// Ports that were added to np.localPods before shouldn't be passed here again
+func (oc *Controller) updateDenyPortGroupsAdd(np *networkPolicy, portNamesToUUIDs map[string]string) error {
+	var err error
+	var ops []ovsdb.Operation
+	ingressDenyPGName := defaultDenyPortGroupName(np.namespace, ingressDefaultDenySuffix)
+	egressDenyPGName := defaultDenyPortGroupName(np.namespace, egressDefaultDenySuffix)
+
+	pgKey := np.namespace
+	// this lock guarantees that sharedPortGroup counters and np.localPods will be updated together
+	// with adding port to port group atomically
+	// need to use write lock, because if db ops fail after counters were updated,
+	// other threads could've already read updated counter and think that port is already added to port group,
+	// but in reality it is not
+	oc.sharedNetpolPortGroups.LockKey(pgKey)
+	defer oc.sharedNetpolPortGroups.UnlockKey(pgKey)
+	sharedPGs, ok := oc.sharedNetpolPortGroups.Load(pgKey)
+	if !ok {
+		// Port group doesn't exist
+		return fmt.Errorf("port groups for ns %s don't exist", np.namespace)
+	}
+
+	ingressDenyPorts, egressDenyPorts := sharedPGs.updateDefaultDenyPortsCountersAdd(np, portNamesToUUIDs)
+	// counters were updated, update back to initial values on error
+	defer func() {
 		if err != nil {
-			klog.Errorf(err.Error())
-			return
+			sharedPGs.updateDefaultDenyPortsCountersDel(np, portNamesToUUIDs)
 		}
+	}()
 
-		// If we never saw this pod, short-circuit
-		if _, ok := np.localPods.LoadAndDelete(logicalPort); !ok {
-			continue
+	if len(ingressDenyPorts) != 0 || len(egressDenyPorts) != 0 {
+		// db changes required
+		ops, err = libovsdbops.AddPortsToPortGroupOps(oc.nbClient, ops, ingressDenyPGName, ingressDenyPorts...)
+		if err != nil {
+			return fmt.Errorf("unable to get add ports to %s port group ops: %v", ingressDenyPGName, err)
+		}
+
+		ops, err = libovsdbops.AddPortsToPortGroupOps(oc.nbClient, ops, egressDenyPGName, egressDenyPorts...)
+		if err != nil {
+			return fmt.Errorf("unable to get add ports to %s port group ops: %v", egressDenyPGName, err)
 		}
 
-		policyPortsInfo = append(policyPortsInfo, portInfo)
-		policyPorts = append(policyPorts, portInfo.uuid)
+		_, err = libovsdbops.TransactAndCheck(oc.nbClient, ops)
+		if err != nil {
+			return fmt.Errorf("unable to transact add ports to default deny port groups: %v", err)
+		}
 	}
+	return nil
+}
 
-	ingressDenyPorts, egressDenyPorts = oc.localPodDelDefaultDeny(np, policyPortsInfo...)
+// updateDenyPortGroupsDel deletes ports from default deny port groups, and atomically updates
+// 1. oc.lsp[In/E]gressDenyCache counters
+// 2. default deny port groups in the db
+// 3. np.localPods
+// Ports that were deleted from np.localPods before shouldn't be passed here again
+func (oc *Controller) updateDenyPortGroupsDel(np *networkPolicy, portNamesToUUIDs map[string]string, useLocalPods bool) error {
```

Comment:
explain what useLocalPods is for in the function text

Reply:
actually cant we just say if the len(portNamesToUUIDs) == 0, then we assume we want to delete all pods from this port group?

Reply:
len(portNamesToUUIDs) == 0 means we don't want to delete anything

Reply:
but we could change that? why would we call the function if there is nothing to delete?

Reply:
why use "special values" that have different meaning from the regular values for the same argument, if I can just explicitly pass the flag? I am not a fan of "this is a list of ports to delete, but if it's empty it means delete all localPods" - that sounds error-prone

Reply:
ok thats fine to keep as is